File: typescript/tests/verification/incidence-matrix.test.ts
import { describe, it, expect } from 'vitest';
import { IncidenceMatrix } from '../../src/verification/encoding/incidence-matrix.js';
import { flatten } from '../../src/verification/encoding/net-flattener.js';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place } from '../../src/core/place.js';
import { one } from '../../src/core/in.js';
import { outPlace } from '../../src/core/out.js';

describe('IncidenceMatrix', () => {
  it('simple net incidence matrix correct', () => {
    const pA = place('A');
    const pB = place('B');
    const t = Transition.builder('T')
      .inputs(one(pA))
      .outputs(outPlace(pB))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);
    const matrix = IncidenceMatrix.from(flatNet);

    expect(matrix.numTransitions()).toBe(1);
    expect(matrix.numPlaces()).toBe(2);

    const idxA = flatNet.placeIndex.get('A')!;
    const idxB = flatNet.placeIndex.get('B')!;

    // Pre: A consumed (1), B not consumed (0)
    expect(matrix.pre()[0]![idxA]).toBe(1);
    expect(matrix.pre()[0]![idxB]).toBe(0);

    // Post: A not produced (0), B produced (1)
    expect(matrix.post()[0]![idxA]).toBe(0);
    expect(matrix.post()[0]![idxB]).toBe(1);

    // Incidence: A = 0-1 = -1, B = 1-0 = +1
    expect(matrix.incidence()[0]![idxA]).toBe(-1);
    expect(matrix.incidence()[0]![idxB]).toBe(1);
  });

  it('transposed incidence dimensions correct', () => {
    const pA = place('A');
    const pB = place('B');
    const pC = place('C');
    const t1 = Transition.builder('T1')
      .inputs(one(pA))
      .outputs(outPlace(pB))
      .build();
    const t2 = Transition.builder('T2')
      .inputs(one(pB))
      .outputs(outPlace(pC))
      .build();
    const net = PetriNet.builder('N').transitions(t1, t2).build();
    const flatNet = flatten(net);
    const matrix = IncidenceMatrix.from(flatNet);

    const ct = matrix.transposedIncidence();
    // C^T should be P×T
    expect(ct.length).toBe(matrix.numPlaces());
    expect(ct[0]!.length).toBe(matrix.numTransitions());
  });

  it('circular net incidence sums to zero per place', () => {
    const pA = place('A');
    const pB = place('B');
    const t1 = Transition.builder('T1')
      .inputs(one(pA))
      .outputs(outPlace(pB))
      .build();
    const t2 = Transition.builder('T2')
      .inputs(one(pB))
      .outputs(outPlace(pA))
      .build();
    const net = PetriNet.builder('N').transitions(t1, t2).build();
    const flatNet = flatten(net);
    const matrix = IncidenceMatrix.from(flatNet);

    const ct = matrix.transposedIncidence();
    // For each place (row), sum of incidence across transitions should be 0
    for (let p = 0; p < matrix.numPlaces(); p++) {
      let sum = 0;
      for (let t = 0; t < matrix.numTransitions(); t++) {
        sum += ct[p]![t]!;
      }
      expect(sum).toBe(0);
    }
  });

  it('pipeline net incidence matrix', () => {
    const pA = place('A');
    const pB = place('B');
    const pC = place('C');
    const t1 = Transition.builder('T1')
      .inputs(one(pA))
      .outputs(outPlace(pB))
      .build();
    const t2 = Transition.builder('T2')
      .inputs(one(pB))
      .outputs(outPlace(pC))
      .build();
    const net = PetriNet.builder('N').transitions(t1, t2).build();
    const flatNet = flatten(net);
    const matrix = IncidenceMatrix.from(flatNet);

    expect(matrix.numTransitions()).toBe(2);
    expect(matrix.numPlaces()).toBe(3);
  });
});

File: java/src/test/java/org/libpetri/smt/encoding/IncidenceMatrixTest.java
package org.libpetri.smt.encoding;

import org.libpetri.analysis.EnvironmentAnalysisMode;
import org.libpetri.core.*;
import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class IncidenceMatrixTest {

    @Test
    void simpleNet_incidenceMatrixCorrect() {
        var p1 = Place.of("A", String.class);
        var p2 = Place.of("B", String.class);

        var t = Transition.builder("T")
            .inputs(In.one(p1))
            .outputs(Out.place(p2))
            .build();

        var net = PetriNet.builder("Test").transitions(t).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());
        var matrix = IncidenceMatrix.from(flatNet);

        int idxA = flatNet.indexOf(p1);
        int idxB = flatNet.indexOf(p2);

        assertEquals(1, matrix.numTransitions());
        assertEquals(2, matrix.numPlaces());

        // T consumes from A, produces to B
        assertEquals(1, matrix.pre()[0][idxA]);
        assertEquals(0, matrix.pre()[0][idxB]);
        assertEquals(0, matrix.post()[0][idxA]);
        assertEquals(1, matrix.post()[0][idxB]);
        assertEquals(-1, matrix.incidence()[0][idxA]);
        assertEquals(1, matrix.incidence()[0][idxB]);
    }

    @Test
    void transposedIncidence_dimensionsCorrect() {
        var p1 = Place.of("A", String.class);
        var p2 = Place.of("B", String.class);
        var p3 = Place.of("C", String.class);

        var t1 = Transition.builder("T1")
            .inputs(In.one(p1))
            .outputs(Out.place(p2))
            .build();
        var t2 = Transition.builder("T2")
            .inputs(In.one(p2))
            .outputs(Out.place(p3))
            .build();

        var net = PetriNet.builder("Test").transitions(t1, t2).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());
        var matrix = IncidenceMatrix.from(flatNet);

        int[][] ct = matrix.transposedIncidence();
        assertEquals(3, ct.length, "C^T should have P rows");
        assertEquals(2, ct[0].length, "C^T should have T columns");
    }

    @Test
    void circularNet_incidenceMatrixSumsToZeroPerRow() {
        // A -> B -> A (circular)
        var pA = Place.of("A", String.class);
        var pB = Place.of("B", String.class);

        var t1 = Transition.builder("Forward")
            .inputs(In.one(pA))
            .outputs(Out.place(pB))
            .build();
        var t2 = Transition.builder("Back")
            .inputs(In.one(pB))
            .outputs(Out.place(pA))
            .build();

        var net = PetriNet.builder("Circular").transitions(t1, t2).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());
        var matrix = IncidenceMatrix.from(flatNet);

        // For a circular net, the sum of incidence columns (across all transitions)
        // for each place should equal 0 (tokens are conserved)
        int[][] ct = matrix.transposedIncidence();
        for (int p = 0; p < matrix.numPlaces(); p++) {
            int sum = 0;
            for (int t = 0; t < matrix.numTransitions(); t++) {
                sum += ct[p][t];
            }
            assertEquals(0, sum, "Place " + p + " should have zero net token change in circular net");
        }
    }
}
